arch/eventloop_lws_http.c

```diff
@@ -0,0 +1,259 @@
+        char *px = buffer + LWS_PRE;
+        int lenx = sizeof(buffer) - LWS_PRE;
+
+        if (lws_http_client_read(wsi, &px, &lenx) < 0)
+            return -1;
+    }
+        return 0; /* don't passthru */
+
+    case LWS_CALLBACK_COMPLETED_CLIENT_HTTP:
+        lwsl_user("LWS_CALLBACK_COMPLETED_CLIENT_HTTP %s\n",
+                  lws_wsi_tag(wsi));
+        break;
+
+    case LWS_CALLBACK_CLOSED_CLIENT_HTTP:
+        lwsl_info("%s: closed: %s\n", __func__, lws_wsi_tag(wsi));
+        break;
+
+    default:
+        break;
+    }
+    return 0;
+}
+/* usually defined as const */
+static struct lws_protocols protocols[] = {
+    { "http", callback_http, 0, 0, 0, NULL, 0},
+    LWS_PROTOCOL_LIST_TERM
+};
+
+struct HTTPConnectionManager {
+    UA_ConnectionManager cm;
+    struct lws_context *context;
+};
+
+static UA_StatusCode
+HTTP_eventSourceStart(UA_ConnectionManager *cm) {
+    UA_EventLoop *el = cm->eventSource.eventLoop;
+    if(!el)
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    /* Check the state */
+    if(cm->eventSource.state != UA_EVENTSOURCESTATE_STOPPED) {
+        UA_LOG_ERROR(el->logger, UA_LOGCATEGORY_NETWORK,
+                     "HTTP\t| To start the ConnectionManager, it has to be "
+                     "registered in an EventLoop and not started yet");
+        return UA_STATUSCODE_BADINTERNALERROR;
+    }
+
+    cm->eventSource.state = UA_EVENTSOURCESTATE_STARTED;
+    return UA_STATUSCODE_GOOD;
+}
+
+static void
+HTTP_eventSourceStop(UA_ConnectionManager *cm) {
+    if(cm->eventSource.state == UA_EVENTSOURCESTATE_STOPPING ||
+       cm->eventSource.state == UA_EVENTSOURCESTATE_STOPPED)
+        return;
+
+    UA_LOG_INFO(cm->eventSource.eventLoop->logger, UA_LOGCATEGORY_NETWORK,
+                "HTTP\t| Shutting down the ConnectionManager");
+
+    cm->eventSource.state = UA_EVENTSOURCESTATE_STOPPING;
+
+    /* Stopping and no connection open? */
+    HTTPConnectionManager *mcm = (HTTPConnectionManager*)cm;
+    if(mcm->cm.eventSource.state == UA_EVENTSOURCESTATE_STOPPING) {
+        mcm->cm.eventSource.state = UA_EVENTSOURCESTATE_STOPPED;
+        return;
+    }
+}
+
+static UA_StatusCode
+HTTP_eventSourceDelete(UA_ConnectionManager *cm) {
+    UA_String_clear(&cm->eventSource.name);
+    UA_free(cm);
+    return UA_STATUSCODE_GOOD;
+}
+
+/* The function is not needed for HTTP.
+ * Request can be sent directly with the sendWithConnection function. */
+static UA_StatusCode
+HTTP_openConnection(UA_ConnectionManager *cm, const UA_KeyValueMap *params,
+                    void *application, void *context,
+                    UA_ConnectionManager_connectionCallback connectionCallback) {
+    HTTPConnectionManager *hcm = (HTTPConnectionManager*)cm;
+
+    protocols[0].user = (void*)cm;
+
+    /* for the purpose of debugging */
+    //int logs = LLL_USER | LLL_ERR | LLL_WARN | LLL_NOTICE;
+    lws_set_log_level(0, NULL);
+
+    struct lws_context_creation_info info;
+    void *foreign_loops[1];
```

Comment:
I have not understood this construct.
Why not just say `info.foreign_loops = hcm->cm.eventSource.eventLoop;`?